src/modm/driver/can/mcp2515_impl.hpp

```diff
@@ -186,124 +178,232 @@ modm::Mcp2515<SPI, CS, INT>::setMode(Can::Mode mode)
+template <typename SPI, typename CS, typename INT>
+bool
+modm::Mcp2515<SPI, CS, INT>::sendMessage(const can::Message& message)
+{
+	bool success = true;
+	if (not modm_assert_continue_ignore(txQueue.push(message), "mcp2515.can.tx",
+			"CAN transmit software buffer overflowed!", 1)) {
+		/// buffer full, could not send
+		success = false;
+	}
+	return success;
+}
+
+// ----------------------------------------------------------------------------
+template <typename SPI, typename CS, typename INT>
+modm::ResumableResult<bool>
+modm::Mcp2515<SPI, CS, INT>::mcp2515ReadMessage()
 {
 	using namespace mcp2515;
 
-	uint8_t status = readStatus(RX_STATUS);
-	uint8_t address;
-	if (status & FLAG_RXB0_FULL) {
-		address = READ_RX;			// message in buffer 0
+	RF_BEGIN();
+
+	// read status flag of the device
+	statusBufferR_ = RF_CALL(readStatus(RX_STATUS));
+
+	tempR_ = true;
+	if (statusBufferR_ & FLAG_RXB0_FULL) {
+		addressBufferR_ = READ_RX;			// message in buffer 0
 	}
-	else if (status & FLAG_RXB1_FULL) {
-		address = READ_RX | 0x04;	// message in buffer 1 (RXB1SIDH)
+	else if (statusBufferR_ & FLAG_RXB1_FULL) {
+		addressBufferR_ = READ_RX | 0x04;	// message in buffer 1 (RXB1SIDH)
 	}
 	else {
-		return false;				// Error: no message available
+		tempR_ = false;						// Error: no message available
 	}
 
-	chipSelect.reset();
-	spi.transferBlocking(address);
+	if(tempR_)
+	{
+		RF_WAIT_UNTIL(this->acquireMaster());
+		chipSelect.reset();
+		RF_CALL(spi.transfer(addressBufferR_));
 
-	message.flags.extended = readIdentifier(message.identifier);
-	if (status & FLAG_RTR) {
-		message.flags.rtr = true;
-	}
-	else {
-		message.flags.rtr = false;
-	}
-	message.length = spi.transferBlocking(0xff) & 0x0f;
+		messageBuffer_.flags.extended = RF_CALL(readIdentifier(messageBuffer_.identifier));
+		if (statusBufferR_ & FLAG_RTR) {
+			messageBuffer_.flags.rtr = true;
+		}
+		else {
+			messageBuffer_.flags.rtr = false;
+		}
 
-	for (uint8_t i = 0; i < message.length; ++i) {
-		message.data[i] = spi.transferBlocking(0xff);
-	}
-	chipSelect.set();
+		messageBuffer_.length = RF_CALL(spi.transfer(0xff)) & 0x0f;
 
+		for (j_ = 0; j_ < messageBuffer_.length; ++j_) {
+			data_ = RF_CALL(spi.transfer(0xff));
+			messageBuffer_.data[j_] = data_;
+		}
+		RF_WAIT_UNTIL(this->releaseMaster());
+		chipSelect.set();
+	}
 	// RX0IF or RX1IF respectivly were already cleared automatically by rising CS.
 	// See section 12.4 in datasheet.
 
-	return true;
+	RF_END_RETURN(tempR_);
+}
+
+template <typename SPI, typename CS, typename INT>
+modm::ResumableResult<bool>
+modm::Mcp2515<SPI, CS, INT>::update(){
+	using namespace mcp2515;
+
+	RF_BEGIN();
+
+	// check if the device has received a message(pin = LOW)
+	// if yes: read it and put it into the rxQueue
+	if(!interruptPin.read()){
+		if(RF_CALL(mcp2515ReadMessage()))
+		{
+			if(not modm_assert_continue_ignore(rxQueue.push(messageBuffer_), "mcp2515.can.tx",
+			"CAN transmit software buffer overflowed!", 1)){
+				/// ignore
+			}
```

Comment:
Remove the if statement if the body is empty.

Reply:
well it was there before ... it shows intent .. so i kept it for now